### PR TITLE
Loot part3 - Monster drops loot on the ground

### PR DESCRIPTION
**Pull request by Cr3ativeZen**

### CODE REVIEW

File: apps/gameserver/include/Loot.h [hunk 1]
```diff
@@ -77,3 +78,29 @@ struct ItemGroup : public bango::processor::db_object<ItemGroup>
 };
 
 std::vector<std::uint32_t> GetValuesFromBrackets(bango::processor::lisp::var& param);
+
+
+class Loot : public Character
+{
+    ITEMINFO m_item_info;
```
Comment (lafreak) [inline apps/gameserver/include/Loot.h:85]: ITEMINFO might not have default constructor which initializes the data. This means we have garbage upon Loot creation. 

In Loot constructor we only fill those:
```
    m_item_info.Index = loot_info.m_index;
    m_item_info.Num = loot_info.m_amount;
    m_item_info.Prefix = loot_info.m_prefix;
```

My suggestion is to memset all zeros m_item_info first in constructor. Please look it up and decide what's most clean option.

File: apps/gameserver/src/Loot.cpp [hunk 2]
```diff
@@ -136,4 +139,42 @@ const Group* ItemGroup::RollGroup() const
     }
 
     return group;
+}
+
+Loot::Loot(LootInfo loot_info, int x, int y, int map)
+    : Character(Character::LOOT)
+{
+    AssignNewId();
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:147]: Other Character inherited class do not need to call this in constructor. Please use the same approach.

File: apps/gameserver/src/Loot.cpp [hunk 3]
```diff
@@ -136,4 +139,42 @@ const Group* ItemGroup::RollGroup() const
...
+    SetAppearTime(time::now());
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:148]: We do not need an API to set any time. We can encapsulate the behavior to `ResetAppearTime()` and internal implementation would choose time::now()

Comment (lafreak) [reply apps/gameserver/src/Loot.cpp:148]: It's good to never expose more functionalities than you need

File: apps/gameserver/include/World.h [hunk 4]
```diff
@@ -238,6 +260,10 @@ class World
             case Character::NPC:
                 Get().m_npcs.insert(std::make_pair(entity->GetID(), std::dynamic_pointer_cast<NPC>(entity)));
                 break;
+            case Character::LOOT:
+                Get().m_loots.insert(std::make_pair(entity->GetID(), std::dynamic_pointer_cast<Loot>(entity)));
+                auto loots = Get().m_loots;
```
Comment (lafreak) [inline apps/gameserver/include/World.h:265]: ?

File: apps/gameserver/src/Loot.cpp [hunk 5]
```diff
@@ -1,8 +1,11 @@
 #include "Loot.h"
 
 #include <bango/utils/random.h>
+#include "World.h"
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:4]: Why is World needed?

File: apps/gameserver/src/World.cpp [hunk 6]
```diff
@@ -99,6 +99,14 @@ void WorldMap::Add(Character* entity)
             if (monster->distance(center) <= m_sight)
                 m_on_appear(*(Player*) entity, *monster, false);
         }
+
+        for (auto& p : container->loots())
+        {
+            auto loot = p.second;
+
+            if (loot->distance(center) <= m_sight)
+                m_on_appear(*(Player*) entity, *loot, false);
```
Comment (lafreak) [inline apps/gameserver/src/World.cpp:108]: Missing `m_on_disappear` call for loot
World.cpp L187-L200

File: apps/gameserver/src/Loot.cpp (hunk 2 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Loot.cpp:147]: ups, done

File: apps/gameserver/include/World.h (hunk 4 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/World.h:265]: ups, done

File: apps/gameserver/src/Loot.cpp (hunk 5 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Loot.cpp:4]: forget to remove, done

File: apps/gameserver/src/World.cpp (hunk 6 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/World.cpp:108]: done

File: apps/gameserver/src/Loot.cpp (hunk 3 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Loot.cpp:148]: done

File: apps/gameserver/include/Loot.h (hunk 1 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Loot.h:85]: memset added.

**Review by lafreak — APPROVED** [review]

**Pull request merged**